File: src/_cpp/TowerDefenseBase/scenes/gamegs/HeroIcon.cpp
#include "HeroIcon.h"
#include "configuration.h"
#include "ml/ImageManager.h"
#include "ml/loadxml/xmlProperties.h"
NS_CC_BEGIN


namespace
{
	const std::string ext( ".png" );
	const std::string empty( "" );
};


HeroIcon::HeroIcon()
{
}

HeroIcon::~HeroIcon( )
{
}

bool HeroIcon::init( const std::string & heroname, const ccMenuCallback & callback )
{
	do
	{
		CC_BREAK_IF( !MenuItemImageWithText::initWithNormalImage( empty, empty, empty, empty, empty, callback ) );
		std::string cancel = k::resourceGameSceneFolder + "icon_x_10.png";
		std::string normal = k::resourceGameSceneFolder + heroname + "_2" + ext;
		std::string selected = k::resourceGameSceneFolder + heroname + "_3" + ext;
		std::string disabled = k::resourceGameSceneFolder + heroname + "_1" + ext;

		_unselectedHero = normal;
		_selectedHero = selected;
		
		std::string timer = k::resourceGameSceneFolder + "hero_progressbar1" + ext;
		std::string timer_bg = k::resourceGameSceneFolder + "hero_progressbar2" + ext;

		setImageNormal( normal );
		setImageDisabled( disabled );

		auto progress = Node::create();
		Sprite * image = ImageManager::sprite( timer );
		CC_BREAK_IF( !image );
		_timer = ProgressTimer::create( image );
		CC_BREAK_IF( !_timer );

		image->setAnchorPoint( Point::ANCHOR_BOTTOM_LEFT );
		_timer->setType( ProgressTimer::Type::BAR );
		_timer->setMidpoint( Point( 0.0f, 0.5f ) );
		_timer->setBarChangeRate( Point( 1, 0 ) );
		_timer->setPercentage( 0 );
		
		progress->addChild( ImageManager::sprite( timer_bg ) );
		progress->addChild( _timer, 1 );

		addChild( progress );
		Point pos;
		pos.x = getNormalImage()->getContentSize().width / 2;
		pos.y = -image->getContentSize().height * 0.7f;
		progress->setPosition( pos );
		
		if( 0 )
		{
			_cancelImage = ImageManager::sprite( cancel );
			getNormalImage()->addChild( _cancelImage );
			_cancelImage->setPosition( Point( getNormalImage()->getContentSize() ) - Point( 5, 5 ) );
			_cancelImage->setVisible( false );
			_cancelImage->setScale( 1.5f );
			
			auto s1 = ScaleTo::create( 0.4f, 1.2f );
			auto s2 = ScaleTo::create( 0.4f, 1.0f );
			auto d = DelayTime::create( 1 );
			auto s = Sequence::create( s1, s2, d, nullptr );
			auto action = RepeatForever::create( EaseInOut::create( s, 0.5f ) );
			_cancelImage->runAction( action );
		}


		return true;
	}
	while( false );
	return false;
}

void HeroIcon::setHero( Hero::Pointer hero )
{
	assert( !_hero );
	assert( hero );
	_hero = hero;
	_hero->observerHealth.add( _ID, 
		[this]( float current, float health )
	{
		float percent = current / health * 100.f;
		_timer->setPercentage( percent );

		bool enabled = _hero->current_state().get_name() != Hero::State::state_death;
		unsigned char opacity = enabled ? 255 : 128;
		setEnabled( enabled );
		getNormalImage()->setOpacity( opacity );
		_timer->setOpacity( opacity );
	} );
}

void HeroIcon::showCancel( bool mode )
{
	if( _cancelImage )
	{
		_cancelImage->setVisible( mode );
	}

	if( _hero )
	{
		_hero->runEvent( mode ? "onselect" : "ondeselect" );
	}

	std::string image = mode ? _selectedHero : _unselectedHero;
	xmlLoader::setProperty( getNormalImage(), xmlLoader::kImage, image );
	//auto frame = ImageManager::shared().spriteFrame( image );
	//if( frame )
	//	static_cast<Sprite*>(getNormalImage())->setSpriteFrame( frame );
}



NS_CC_END

File: src/_eu/scenes/gamegs/HeroIcon.js
/******************************************************************************
 * Copyright 2014-2016 Vladimir Tolmachev
 * Copyright 2016 Visionarity AG
 * Vladimir Tolmachev and Visionarity AG have unlimited commercial
 * licenses for commercial use and customization
 *
 * Author: Vladimir Tolmachev (tolvl@hotmail.com)
 * Ported C++ to Javascript: Visionarity AG / Vladimir Tolmachev
 * Project: Island Defense (JS)
 * If you received the code not from the author, please contact us
 ******************************************************************************/

//Define namespace
var EU = EU || {};

EU.HeroIcon = EU.MenuItemImageWithText.extend({
    _hero : null,
    _timer : null,
    _cancelImage : null,
    _unselectedHero : null,
    _selectedHero : null,

    ctor:function( heroname, callback, target )
    {
        this._super( "", "", "", "", "", callback, target);
        var normal = EU.k.resourceGameSceneFolder + heroname + "_2" + ".png";
        var selected = EU.k.resourceGameSceneFolder + heroname + "_3" + ".png";
        var disabled = EU.k.resourceGameSceneFolder + heroname + "_1" + ".png";

        this._unselectedHero = normal;
        this._selectedHero = selected;

        var timer = EU.k.resourceGameSceneFolder + "hero_progressbar1" + ".png";
        var timer_bg = EU.k.resourceGameSceneFolder + "hero_progressbar2" + ".png";

        this.setImageNormal( normal );
        this.setImageDisabled( disabled );

        var progress = Node.create();
        var image = EU.ImageManager.sprite( timer );
        this._timer = new cc.ProgressTimer( image );

        image.setAnchorPoint( cc.p(0,0) );
        this._timer.setType( cc.ProgressTimer.TYPE_BAR );
        this._timer.setMidpoint( cc.p( 0.0, 0.5 ) );
        this._timer.setBarChangeRate( cc.p( 1, 0 ) );
        this._timer.setPercentage( 0 );

        progress.addChild( EU.ImageManager.sprite( timer_bg ) );
        progress.addChild( this._timer, 1 );

        this.addChild( progress );
        var pos = cc.p(0,0);
        pos.x = this.getNormalImage().getContentSize().width / 2;
        pos.y = -image.getContentSize().height * 0.7;
        progress.setPosition( pos );
    },

    setHero: function( hero )
    {
        EU.assert( !this._hero );
        EU.assert( hero );
        this._hero = hero;
        this._hero.observerHealth.add( __instanceId, this.dispatchHealth.bind(this) );
    },
    dispatchHealth:function(current, health){
        var percent = current / health * 100;
        this._timer.setPercentage( percent );
    
        var enabled = this._hero.current_state().get_name() != Hero.State.state_death;
        var opacity = enabled ? 255 : 128;
        this.setEnabled( enabled );
        this.getNormalImage().setOpacity( opacity );
        this._timer.setOpacity( opacity );
    },
    showCancel: function( mode )
    {
        if( this._cancelImage )
        {
            this._cancelImage.setVisible( mode );
        }

        if( this._hero )
        {
            this._hero.runEvent( mode ? "onselect" : "ondeselect" );
        }

        var image = mode ? this._selectedHero : this._unselectedHero;
        EU.xmlLoader.setProperty_int( this.getNormalImage(), EU.xmlKey.Image.int, image );
    }
});